homeassistant/components/template/binary_sensor.py

```diff
@@ -301,9 +307,33 @@ def __init__(
                 self._parse_result.add(key)
 
         self._delay_cancel: CALLBACK_TYPE | None = None
-        self._auto_off_cancel = None
+        self._auto_off_cancel: CALLBACK_TYPE | None = None
+        self._auto_off_time: datetime | None = None
         self._state: bool | None = None
 
+    async def async_added_to_hass(self) -> None:
+        """Restore last state."""
+        await super().async_added_to_hass()
+        if (
+            (last_state := await self.async_get_last_state()) is not None
+            and (extra_data := await self.async_get_last_binary_sensor_data())
+            is not None
+            and last_state.state not in (STATE_UNKNOWN, STATE_UNAVAILABLE)
+        ):
+            self._state = last_state.state == STATE_ON
```

Comment:
Maybe we should only restore state if either of `CONF_AUTO_OFF`; `CONF_DELAY_ON`; `CONF_DELAY_OFF` are set?

Reply:
IMO restart/reload should not change the state of a trigger entity, only the trigger should. So they all should restore state to ensure that they always remain in whatever state was determined by the last trigger until the next trigger.

I am fine with adding a config option to control this behavior though in case there are cases I'm not considering where not restoring state is desirable. I don't know if I agree that these config options make them more restorable. If anything `auto_off` sensors are the only ones that we could maybe skip restoring and just default to `off` for the reasons suggested by the author in the linked issue. But I think what you're doing here with restoring both the state and the timer sounds much better, I really like it.